[CI] PyTorchModelIT testEvaluateWithMinimalTimeout failing

**Build scan:**
https://gradle-enterprise.elastic.co/s/6wkkmz5vkjula/tests/:x-pack:plugin:ml:qa:native-multi-node-tests:javaRestTest/org.elasticsearch.xpack.ml.integration.PyTorchModelIT/testEvaluateWithMinimalTimeout

**Reproduction line:**
`./gradlew ':x-pack:plugin:ml:qa:native-multi-node-tests:javaRestTest' --tests "org.elasticsearch.xpack.ml.integration.PyTorchModelIT.testEvaluateWithMinimalTimeout" -Dtests.seed=2B611D44EF33CA82 -Dtests.locale=ga-IE -Dtests.timezone=Etc/Universal -Druntime.java=17`

**Applicable branches:**
master

**Reproduces locally?:**
No

**Failure history:**
https://gradle-enterprise.elastic.co/scans/tests?tests.container=org.elasticsearch.xpack.ml.integration.PyTorchModelIT&tests.test=testEvaluateWithMinimalTimeout

**Failure excerpt:**
```
java.lang.AssertionError: 3 active tasks found:
cluster:monitor/xpack/ml/inference/infer                    CJ34ywReS9aVivmGA_PdHA:82102 -                            transport 1654558842528 23:40:42 1.5m        127.0.0.1 javaRestTest-2 
cluster:monitor/xpack/ml/trained_models/deployment/infer    CJ34ywReS9aVivmGA_PdHA:82103 CJ34ywReS9aVivmGA_PdHA:82102 transport 1654558842530 23:40:42 1.5m        127.0.0.1 javaRestTest-2 
cluster:monitor/xpack/ml/trained_models/deployment/infer[n] CJ34ywReS9aVivmGA_PdHA:82104 CJ34ywReS9aVivmGA_PdHA:82103 direct    1654558842531 23:40:42 1.5m        127.0.0.1 javaRestTest-2 
 expected:<0> but was:<3>

  at __randomizedtesting.SeedInfo.seed([2B611D44EF33CA82:CC1762B640F8F442]:0)
  at org.junit.Assert.fail(Assert.java:88)
  at org.junit.Assert.failNotEquals(Assert.java:834)
  at org.junit.Assert.assertEquals(Assert.java:645)
  at org.elasticsearch.test.rest.ESRestTestCase.lambda$waitForPendingTasks$2(ESRestTestCase.java:490)
  at org.elasticsearch.test.ESTestCase.assertBusy(ESTestCase.java:1096)
  at org.elasticsearch.test.rest.ESRestTestCase.waitForPendingTasks(ESRestTestCase.java:463)
  at org.elasticsearch.test.rest.ESRestTestCase.waitForPendingTasks(ESRestTestCase.java:450)
  at org.elasticsearch.xpack.ml.integration.PyTorchModelIT.cleanup(PyTorchModelIT.java:145)
  at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(NativeMethodAccessorImpl.java:-2)
  at jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
  at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
  at java.lang.reflect.Method.invoke(Method.java:568)
  at com.carrotsearch.randomizedtesting.RandomizedRunner.invoke(RandomizedRunner.java:1758)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$10.evaluate(RandomizedRunner.java:1004)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at org.apache.lucene.tests.util.TestRuleSetupTeardownChained$1.evaluate(TestRuleSetupTeardownChained.java:44)
  at org.apache.lucene.tests.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:43)
  at org.apache.lucene.tests.util.TestRuleThreadAndTestName$1.evaluate(TestRuleThreadAndTestName.java:45)
  at org.apache.lucene.tests.util.TestRuleIgnoreAfterMaxFailures$1.evaluate(TestRuleIgnoreAfterMaxFailures.java:60)
  at org.apache.lucene.tests.util.TestRuleMarkFailure$1.evaluate(TestRuleMarkFailure.java:44)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl$StatementRunner.run(ThreadLeakControl.java:375)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl.forkTimeoutingTask(ThreadLeakControl.java:824)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl$3.evaluate(ThreadLeakControl.java:475)
  at com.carrotsearch.randomizedtesting.RandomizedRunner.runSingleTest(RandomizedRunner.java:955)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$5.evaluate(RandomizedRunner.java:840)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$6.evaluate(RandomizedRunner.java:891)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$7.evaluate(RandomizedRunner.java:902)
  at org.apache.lucene.tests.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:43)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at org.apache.lucene.tests.util.TestRuleStoreClassName$1.evaluate(TestRuleStoreClassName.java:38)
  at com.carrotsearch.randomizedtesting.rules.NoShadowingOrOverridesOnMethodsRule$1.evaluate(NoShadowingOrOverridesOnMethodsRule.java:40)
  at com.carrotsearch.randomizedtesting.rules.NoShadowingOrOverridesOnMethodsRule$1.evaluate(NoShadowingOrOverridesOnMethodsRule.java:40)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at org.apache.lucene.tests.util.TestRuleAssertionsRequired$1.evaluate(TestRuleAssertionsRequired.java:53)
  at org.apache.lucene.tests.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:43)
  at org.apache.lucene.tests.util.TestRuleMarkFailure$1.evaluate(TestRuleMarkFailure.java:44)
  at org.apache.lucene.tests.util.TestRuleIgnoreAfterMaxFailures$1.evaluate(TestRuleIgnoreAfterMaxFailures.java:60)
  at org.apache.lucene.tests.util.TestRuleIgnoreTestSuites$1.evaluate(TestRuleIgnoreTestSuites.java:47)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl$StatementRunner.run(ThreadLeakControl.java:375)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl.lambda$forkTimeoutingTask$0(ThreadLeakControl.java:831)
  at java.lang.Thread.run(Thread.java:833)

```

--- a/docs/changelog/87533.yaml
+++ b/docs/changelog/87533.yaml
@@ -1,0 +1,5 @@
+pr: 87533
+summary: Fixes inference timeout handling bug that throws unexpected `NullPointerException`
+area: Machine Learning
+type: bug
+issues: []

--- a/x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/deployment/AbstractPyTorchAction.java
+++ b/x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/deployment/AbstractPyTorchAction.java
@@ -10,28 +10,28 @@ package org.elasticsearch.xpack.ml.inference.deployment;
 import org.apache.logging.log4j.Logger;
 import org.elasticsearch.ElasticsearchStatusException;
 import org.elasticsearch.action.ActionListener;
-import org.elasticsearch.common.util.concurrent.AbstractRunnable;
 import org.elasticsearch.core.TimeValue;
 import org.elasticsearch.rest.RestStatus;
 import org.elasticsearch.threadpool.Scheduler;
 import org.elasticsearch.threadpool.ThreadPool;
 import org.elasticsearch.xpack.core.ml.utils.ExceptionsHelper;
 import org.elasticsearch.xpack.ml.MachineLearning;
+import org.elasticsearch.xpack.ml.job.process.AbstractInitializableRunnable;
 
 import java.util.concurrent.atomic.AtomicBoolean;
 
 import static org.elasticsearch.core.Strings.format;
 
-abstract class AbstractPyTorchAction<T> extends AbstractRunnable {
+abstract class AbstractPyTorchAction<T> extends AbstractInitializableRunnable {
 
     private final String modelId;
     private final long requestId;
     private final TimeValue timeout;
-    private final Scheduler.Cancellable timeoutHandler;
+    private Scheduler.Cancellable timeoutHandler;
     private final DeploymentManager.ProcessContext processContext;
     private final AtomicBoolean notified = new AtomicBoolean();
-
     private final ActionListener<T> listener;
+    private final ThreadPool threadPool;
 
     protected AbstractPyTorchAction(
         String modelId,
@@ -41,16 +41,23 @@ abstract class AbstractPyTorchAction<T> extends AbstractRunnable {
         ThreadPool threadPool,
         ActionListener<T> listener
     ) {
-        this.modelId = modelId;
+        this.modelId = ExceptionsHelper.requireNonNull(modelId, "modelId");
         this.requestId = requestId;
-        this.timeout = timeout;
-        this.timeoutHandler = threadPool.schedule(
-            this::onTimeout,
-            ExceptionsHelper.requireNonNull(timeout, "timeout"),
-            MachineLearning.UTILITY_THREAD_POOL_NAME
-        );
-        this.processContext = processContext;
-        this.listener = listener;
+        this.timeout = ExceptionsHelper.requireNonNull(timeout, "timeout");
+        this.processContext = ExceptionsHelper.requireNonNull(processContext, "processContext");
+        this.listener = ExceptionsHelper.requireNonNull(listener, "listener");
+        this.threadPool = ExceptionsHelper.requireNonNull(threadPool, "threadPool");
+    }
+
+    /**
+     * Needs to be called after construction. This init starts the timeout handler and needs to be called before added to the executor for
+     * scheduled work.
+     */
+    @Override
+    public final void init() {
+        if (this.timeoutHandler == null) {
+            this.timeoutHandler = threadPool.schedule(this::onTimeout, timeout, MachineLearning.UTILITY_THREAD_POOL_NAME);
+        }
     }
 
     void onTimeout() {
@@ -66,7 +73,11 @@ abstract class AbstractPyTorchAction<T> extends AbstractRunnable {
     }
 
     void onSuccess(T result) {
-        timeoutHandler.cancel();
+        if (timeoutHandler != null) {
+            timeoutHandler.cancel();
+        } else {
+            assert false : "init() not called, timeout handler unexpectedly null";
+        }
         if (notified.compareAndSet(false, true)) {
             listener.onResponse(result);
             return;
@@ -75,8 +86,18 @@ abstract class AbstractPyTorchAction<T> extends AbstractRunnable {
     }
 
     @Override
+    public void onRejection(Exception e) {
+        super.onRejection(e);
+        processContext.getRejectedExecutionCount().incrementAndGet();
+    }
+
+    @Override
     public void onFailure(Exception e) {
-        timeoutHandler.cancel();
+        if (timeoutHandler != null) {
+            timeoutHandler.cancel();
+        } else {
+            assert false : "init() not called, timeout handler unexpectedly null";
+        }
         if (notified.compareAndSet(false, true)) {
             processContext.getResultProcessor().ignoreResponseWithoutNotifying(String.valueOf(requestId));
             listener.onFailure(e);

--- a/x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/pytorch/PriorityProcessWorkerExecutorService.java
+++ b/x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/pytorch/PriorityProcessWorkerExecutorService.java
@@ -7,10 +7,10 @@
 
 package org.elasticsearch.xpack.ml.inference.pytorch;
 
-import org.elasticsearch.common.util.concurrent.AbstractRunnable;
 import org.elasticsearch.common.util.concurrent.EsRejectedExecutionException;
 import org.elasticsearch.common.util.concurrent.ThreadContext;
 import org.elasticsearch.core.SuppressForbidden;
+import org.elasticsearch.xpack.ml.job.process.AbstractInitializableRunnable;
 import org.elasticsearch.xpack.ml.job.process.AbstractProcessWorkerExecutorService;
 
 import java.util.concurrent.PriorityBlockingQueue;
@@ -35,7 +35,7 @@ public class PriorityProcessWorkerExecutorService extends AbstractProcessWorkerE
      * A Runnable sorted first by RequestPriority then a tie breaker which in
      * most cases will be the insertion order
      */
-    public static record OrderedRunnable(RequestPriority priority, long tieBreaker, Runnable runnable)
+    public record OrderedRunnable(RequestPriority priority, long tieBreaker, Runnable runnable)
         implements
             Comparable<OrderedRunnable>,
             Runnable {
@@ -76,7 +76,8 @@ public class PriorityProcessWorkerExecutorService extends AbstractProcessWorkerE
      * @param priority Request priority
      * @param tieBreaker For sorting requests of equal priority
      */
-    public synchronized void executeWithPriority(AbstractRunnable command, RequestPriority priority, long tieBreaker) {
+    public synchronized void executeWithPriority(AbstractInitializableRunnable command, RequestPriority priority, long tieBreaker) {
+        command.init();
         if (isShutdown()) {
             EsRejectedExecutionException rejected = new EsRejectedExecutionException(processName + " worker service has shutdown", true);
             command.onRejection(rejected);

--- a/x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/job/process/AbstractInitializableRunnable.java
+++ b/x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/job/process/AbstractInitializableRunnable.java
@@ -1,0 +1,17 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.ml.job.process;
+
+import org.elasticsearch.common.util.concurrent.AbstractRunnable;
+
+/**
+ * Abstract runnable that has an `init` function that can be called when passed to an executor
+ */
+public abstract class AbstractInitializableRunnable extends AbstractRunnable {
+    public abstract void init();
+}

--- a/x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/job/process/ProcessWorkerExecutorService.java
+++ b/x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/job/process/ProcessWorkerExecutorService.java
@@ -33,6 +33,9 @@ public class ProcessWorkerExecutorService extends AbstractProcessWorkerExecutorS
 
     @Override
     public synchronized void execute(Runnable command) {
+        if (command instanceof AbstractInitializableRunnable initializableRunnable) {
+            initializableRunnable.init();
+        }
         if (isShutdown()) {
             EsRejectedExecutionException rejected = new EsRejectedExecutionException(processName + " worker service has shutdown", true);
             if (command instanceof AbstractRunnable runnable) {

--- a/x-pack/plugin/ml/src/test/java/org/elasticsearch/xpack/ml/inference/deployment/ControlMessagePyTorchActionTests.java
+++ b/x-pack/plugin/ml/src/test/java/org/elasticsearch/xpack/ml/inference/deployment/ControlMessagePyTorchActionTests.java
@@ -65,7 +65,7 @@ public class ControlMessagePyTorchActionTests extends ESTestCase {
                 tp,
                 listener
             );
-
+            action.init();
             action.onTimeout();
             action.run();
             verify(resultProcessor, times(1)).ignoreResponseWithoutNotifying("1");
@@ -84,6 +84,7 @@ public class ControlMessagePyTorchActionTests extends ESTestCase {
                 tp,
                 listener
             );
+            action.init();
 
             action.onFailure(new IllegalStateException());
             action.run();
@@ -122,6 +123,7 @@ public class ControlMessagePyTorchActionTests extends ESTestCase {
             tp,
             listener
         );
+        action.init();
 
         action.run();
 

--- a/x-pack/plugin/ml/src/test/java/org/elasticsearch/xpack/ml/inference/deployment/DeploymentManagerTests.java
+++ b/x-pack/plugin/ml/src/test/java/org/elasticsearch/xpack/ml/inference/deployment/DeploymentManagerTests.java
@@ -9,7 +9,6 @@ package org.elasticsearch.xpack.ml.inference.deployment;
 
 import org.elasticsearch.action.ActionListener;
 import org.elasticsearch.client.internal.Client;
-import org.elasticsearch.common.util.concurrent.AbstractRunnable;
 import org.elasticsearch.common.util.concurrent.EsRejectedExecutionException;
 import org.elasticsearch.core.TimeValue;
 import org.elasticsearch.test.ESTestCase;
@@ -31,9 +30,6 @@ import static org.elasticsearch.xpack.ml.MachineLearning.JOB_COMMS_THREAD_POOL_N
 import static org.elasticsearch.xpack.ml.MachineLearning.UTILITY_THREAD_POOL_NAME;
 import static org.hamcrest.Matchers.equalTo;
 import static org.hamcrest.Matchers.instanceOf;
-import static org.mockito.ArgumentMatchers.any;
-import static org.mockito.ArgumentMatchers.anyLong;
-import static org.mockito.Mockito.doThrow;
 import static org.mockito.Mockito.mock;
 import static org.mockito.Mockito.when;
 
@@ -74,6 +70,7 @@ public class DeploymentManagerTests extends ESTestCase {
         Long taskId = 1L;
         when(task.getId()).thenReturn(taskId);
         when(task.isStopped()).thenReturn(Boolean.FALSE);
+        when(task.getModelId()).thenReturn("test-rejected");
 
         DeploymentManager deploymentManager = new DeploymentManager(
             mock(Client.class),
@@ -82,9 +79,12 @@ public class DeploymentManagerTests extends ESTestCase {
             mock(PyTorchProcessFactory.class)
         );
 
-        PriorityProcessWorkerExecutorService executorService = mock(PriorityProcessWorkerExecutorService.class);
-        doThrow(new EsRejectedExecutionException("mock executor rejection")).when(executorService)
-            .executeWithPriority(any(AbstractRunnable.class), any(), anyLong());
+        PriorityProcessWorkerExecutorService executorService = new PriorityProcessWorkerExecutorService(
+            tp.getThreadContext(),
+            "test reject",
+            10
+        );
+        executorService.shutdown();
 
         AtomicInteger rejectedCount = new AtomicInteger();
 

--- a/x-pack/plugin/ml/src/test/java/org/elasticsearch/xpack/ml/inference/deployment/InferencePyTorchActionTests.java
+++ b/x-pack/plugin/ml/src/test/java/org/elasticsearch/xpack/ml/inference/deployment/InferencePyTorchActionTests.java
@@ -75,7 +75,7 @@ public class InferencePyTorchActionTests extends ESTestCase {
             tp,
             listener
         );
-
+        action.init();
         action.onSuccess(new WarningInferenceResults("foo"));
         for (int i = 0; i < 10; i++) {
             action.onSuccess(new WarningInferenceResults("foo"));
@@ -95,7 +95,7 @@ public class InferencePyTorchActionTests extends ESTestCase {
             tp,
             listener
         );
-
+        action.init();
         action.onTimeout();
         for (int i = 0; i < 10; i++) {
             action.onSuccess(new WarningInferenceResults("foo"));
@@ -116,7 +116,7 @@ public class InferencePyTorchActionTests extends ESTestCase {
             tp,
             listener
         );
-
+        action.init();
         action.onFailure(new Exception("bar"));
         for (int i = 0; i < 10; i++) {
             action.onSuccess(new WarningInferenceResults("foo"));
@@ -146,7 +146,7 @@ public class InferencePyTorchActionTests extends ESTestCase {
                 tp,
                 listener
             );
-
+            action.init();
             action.onTimeout();
             action.run();
             verify(resultProcessor, times(1)).ignoreResponseWithoutNotifying("1");
@@ -163,7 +163,7 @@ public class InferencePyTorchActionTests extends ESTestCase {
                 tp,
                 listener
             );
-
+            action.init();
             action.onFailure(new IllegalStateException());
             action.run();
             verify(resultProcessor, never()).registerRequest(anyString(), any());

--- a/x-pack/plugin/ml/src/test/java/org/elasticsearch/xpack/ml/inference/pytorch/PriorityProcessWorkerExecutorServiceTests.java
+++ b/x-pack/plugin/ml/src/test/java/org/elasticsearch/xpack/ml/inference/pytorch/PriorityProcessWorkerExecutorServiceTests.java
@@ -7,10 +7,10 @@
 
 package org.elasticsearch.xpack.ml.inference.pytorch;
 
-import org.elasticsearch.common.util.concurrent.AbstractRunnable;
 import org.elasticsearch.test.ESTestCase;
 import org.elasticsearch.threadpool.TestThreadPool;
 import org.elasticsearch.threadpool.ThreadPool;
+import org.elasticsearch.xpack.ml.job.process.AbstractInitializableRunnable;
 import org.junit.After;
 
 import java.util.concurrent.atomic.AtomicInteger;
@@ -39,7 +39,12 @@ public class PriorityProcessWorkerExecutorServiceTests extends ESTestCase {
         var r3 = new RunOrderValidator(3, counter);
         executor.executeWithPriority(r3, RequestPriority.NORMAL, 101L);
 
+        assertTrue(r1.initialized);
+        assertTrue(r2.initialized);
+        assertTrue(r3.initialized);
         assertTrue(r3.hasBeenRejected);
+        assertFalse(r1.hasBeenRejected);
+        assertFalse(r2.hasBeenRejected);
     }
 
     public void testQueueCapacityReached_HighestPriority() {
@@ -56,8 +61,11 @@ public class PriorityProcessWorkerExecutorServiceTests extends ESTestCase {
         var r5 = new RunOrderValidator(5, counter);
         executor.executeWithPriority(r5, RequestPriority.NORMAL, 105L);
 
+        assertTrue(r3.initialized);
         assertTrue(r3.hasBeenRejected);
+        assertTrue(highestPriorityAlwaysAccepted.initialized);
         assertFalse(highestPriorityAlwaysAccepted.hasBeenRejected);
+        assertTrue(r5.initialized);
         assertTrue(r5.hasBeenRejected);
     }
 
@@ -77,6 +85,10 @@ public class PriorityProcessWorkerExecutorServiceTests extends ESTestCase {
         executor.executeWithPriority(new ShutdownExecutorRunnable(executor), RequestPriority.NORMAL, 10000L);
 
         executor.start();
+
+        assertTrue(r1.initialized);
+        assertTrue(r2.initialized);
+        assertTrue(r3.initialized);
 
         assertTrue(r1.hasBeenRun);
         assertTrue(r2.hasBeenRun);
@@ -114,16 +126,22 @@ public class PriorityProcessWorkerExecutorServiceTests extends ESTestCase {
         );
     }
 
-    private static class RunOrderValidator extends AbstractRunnable {
+    private static class RunOrderValidator extends AbstractInitializableRunnable {
 
         private boolean hasBeenRun = false;
         private boolean hasBeenRejected = false;
         private final int expectedOrder;
         private final AtomicInteger counter;
+        private boolean initialized = false;
 
         RunOrderValidator(int expectedOrder, AtomicInteger counter) {
             this.expectedOrder = expectedOrder;
             this.counter = counter;
+        }
+
+        @Override
+        public void init() {
+            initialized = true;
         }
 
         @Override
@@ -143,7 +161,7 @@ public class PriorityProcessWorkerExecutorServiceTests extends ESTestCase {
         }
     }
 
-    private static class ShutdownExecutorRunnable extends AbstractRunnable {
+    private static class ShutdownExecutorRunnable extends AbstractInitializableRunnable {
 
         PriorityProcessWorkerExecutorService executor;
 
@@ -162,5 +180,9 @@ public class PriorityProcessWorkerExecutorServiceTests extends ESTestCase {
             executor.shutdown();
         }
 
+        @Override
+        public void init() {
+            // do nothing
+        }
     }
 }

--- a/x-pack/plugin/ml/src/test/java/org/elasticsearch/xpack/ml/job/process/ProcessWorkerExecutorServiceTests.java
+++ b/x-pack/plugin/ml/src/test/java/org/elasticsearch/xpack/ml/job/process/ProcessWorkerExecutorServiceTests.java
@@ -42,11 +42,17 @@ public class ProcessWorkerExecutorServiceTests extends ESTestCase {
         threadPool.generic().execute(executor::start);
         executor.shutdown();
         AtomicBoolean rejected = new AtomicBoolean(false);
-        executor.execute(new AbstractRunnable() {
+        AtomicBoolean initialized = new AtomicBoolean(false);
+        executor.execute(new AbstractInitializableRunnable() {
             @Override
             public void onRejection(Exception e) {
                 assertThat(e, isA(EsRejectedExecutionException.class));
                 rejected.set(true);
+            }
+
+            @Override
+            public void init() {
+                initialized.set(true);
             }
 
             @Override
@@ -61,6 +67,7 @@ public class ProcessWorkerExecutorServiceTests extends ESTestCase {
         });
 
         assertTrue(rejected.get());
+        assertTrue(initialized.get());
     }
 
     public void testAutodetectWorkerExecutorService_TasksNotExecutedCallHandlerOnShutdown() throws Exception {

